#### Case Study: Product Sales Performance




In [1]:
# 001.
# Mean Calculation

# Data for product sales
unitsSold = {
    "Product A": [150, 200, 180, 220],
    "Product B": [120, 140, 160, 190],
    "Product C": [180, 210, 230, 240],
    "Product D": [100, 130, 140, 170],
    "Product E": [190, 220, 210, 250],
}

#Calculate the mean units sold for Product A
productA_sales = unitsSold["Product A"]
meanA_unitsSold = sum(productA_sales)/len(productA_sales)

#Print result
print(f"Product A Mean number of units sold is: {meanA_unitsSold}")


Product A Mean number of units sold is: 187.5


In [3]:
# Data for product Revenue
revenue = {
    "Product A": [3000, 4500, 4200, 5000],
    "Product B": [2400, 3100, 3400, 4100],
    "Product C": [4500, 5000, 5200, 5400],
    "Product D": [2200, 2700, 3100, 3500],
    "Product E": [5000, 5500, 5600, 6000],
}

#calculate mean revenue for Product C
productC_revenue = revenue["Product C"]
meanC_revenue = sum(productC_revenue)/ len(productC_revenue)

print(f"Product C mean revenue is : {meanC_revenue}")


Product C mean revenue is : 5025.0


In [12]:
customerSat = {
    "Product A": [8],
    "Product B": [7],
    "Product C": [9],
    "Product D": [6],
    "Product E": [10],
}

#calculate mean customer satisfaction rating for all products.
customerSatoverall = customerSat.values()

ratingsAll = []
for ratings in customerSat.values():
    ratingsAll.extend(ratings)
meanCustomerSat = sum(ratingsAll)/len(ratingsAll)

print(meanCustomerSat)

#print(f"Mean customer satisfaction rating is : {meanCustomerSat}")


8.0


In [6]:
#Calculate Median for units sold for Product D across the four quarters.
#4. What is the median number of units sold for Product D across the four quarters?

unitsSoldproductD = [100, 130, 140, 170]

# sort the list
unitsSoldproductD.sort()

#calculate the median
if len(unitsSoldproductD) % 2 == 0:
    #Even number of elements, median is the average of the middle two elements.
    median = (unitsSoldproductD[int(len(unitsSoldproductD)/2)] + unitsSoldproductD[int(len(unitsSoldproductD)/2 -1)]) / 2 
else:
    #Odd number of elements, median is the middle element
    median = unitsSoldproductD[int(len(unitsSoldproductD)/2)]

print(f'The median number of units sold for Product D is: {median}')

The median umber of units sold for Product D is: 135.0


In [7]:
#5. Determine the median revenue for all products in Q2. 

Q2revenueAllProducts = [200, 140, 210, 130, 220]

Q2revenueAllProducts.sort()

#calculate the median
if len(Q2revenueAllProducts) % 2 == 0:
    #Even number of elements, median is the average of the middle two elements.
    median = (Q2revenueAllProducts[int(len(Q2revenueAllProducts)/2)] + Q2revenueAllProducts[int(len(Q2revenueAllProducts)/2 -1)]) / 2

else:
    #Odd number of elements, median is the middle element.
    median = Q2revenueAllProducts[int(len(Q2revenueAllProducts)/2)]

print(f'The median number of revenue for all Products in Q2 Quarter is: {median}')


The median number of revenue for all Products in Q2 Quarter is: 200


In [8]:
#6. What is the median of the customer satisfaction ratings across all products?
customerSat = [8, 7, 9, 6, 10]

customerSat.sort()

#calculate the median

if len(customerSat) % 2 == 0:
    #Even number of elements, median is the average of the middle two elements.
    median = (customerSat[int(len(customerSat)/2)] + customerSat[int(len(customerSat)/2 -1)]) / 2

else:
    #Odd number of elements, median is the middle element.
    median = customerSat[int(len(customerSat)/2)]

print(f'Them median customer satisfaction rating is: {median}')

Them median customer satisfaction rating is: 8


In [15]:
#Mode Calculation:
#7. Is there any mode in the number of units sold for Product E across the four quaters?

unitsSoldProductE = [190, 220, 210, 250]
from collections import Counter
productE_counts = Counter(unitsSoldProductE)

# Find the element with the highest frequency

mode = productE_counts.most_common(1)[0][0]

print(f'the mode for units sold for Prdoduct E is: {mode}')


the mode for units sold for Prdoduct E is: 190


In [13]:
#Without importing libraries ????? Not sure with this!!!
def find_mode(nums):
    num_counts = {}
    for num in nums:
        num_counts[num] = num_counts.get(num, 0) + 1

    max_count = max(num_counts.values())
    mode = []
    for num, count in num_counts.items():
        if count == max_count:
            mode.append(num)
    
    return mode

mode = find_mode(unitsSoldProductE)

print(f'the mode for units sold for Product E is', mode)

the mode for units sold for Product E is [190, 220, 210, 250]


In [16]:
#8. Find the mode of revenue for Product B over all four quarters.

revenueProductB = [2400, 3100, 3400, 4100]
from collections import Counter
revenueProductB_counts = Counter(revenueProductB)

# Find the element with the highest frequency

mode = revenueProductB_counts.most_common(1)[0][0]

print(f'the mode revenue for Product B is: {mode}')

the mode revenue for Prdoduct B is: 2400


In [3]:
#9. Determine if there is a mode in the customer satisfaction ratings of all products.

customerSat = [8, 7, 9, 6, 10]
from collections import Counter
customerSat_counts = Counter(customerSat)

# Find the element with the highest frequency

mode = customerSat_counts.most_common(1)[0][0]

print(f'the mode for customer rating is: {mode}')

the mode for customer rating is: 8


In [1]:
#Variance Calculation: 
#10. Calculate the variance in units sold for Product A across all quarters.
import numpy as np
unitsSoldProductA = [150, 120, 180, 100, 190]

# Calculate the variance
variance = np.var(unitsSoldProductA)

print("The variance of unitsSoldProductD is:", variance)

The variance of unitsSoldProductD is: 1176.0


In [ ]:
#11. Find the variance in the revenue of Product E across all four quarters.
import numpy as np
revenueProductE = [5000, 5500, 5600, 6000]

#calculate the variance
revEVariance = np.var(revenueProductE)

print(f'The revenue variance for Product E is: {revEVariance}')

The revenue variance for Product E is: 126875.0


### Manual Calculation
- Calculate the mean (average) of the revenues.
- Subtract the mean from each revenue value to find the squared deviations.
- Average the squared deviations to get the variance.

Mean:
        Mean = (5000+5500+5600+6000)/4 = 5525
​       
Squared deviations:
    (5000−5525)² = (−525)² =275625
    (5500−5525)² = (−25)² =625
    (5600−5525)² = (75)² = 5625
    (6000−5525)² = (475)² = 225625

Average of squared deviations (variance):

Variance = (275625 + 625 + 5625 + 225625)/4 = 507500/4 =126875

So the variance in the revenue of Product E is 126875.

In [5]:
# 12. What is the variance in customer satisfaction ratings for all products?
import numpy as np
customerSat = [8, 7, 9, 6, 10]

#calculate customer satisfaction variance
customerSatVariance = np.var(customerSat)

print(f'Customer satisfaction rating variance is : {customerSatVariance}')

Customer satisfaction rating variance is : 2.0


In [6]:
#Standard Deviation Calculation: 
#13. Compute the standard deviation of units sold for Product B.
import numpy as np

unitsSoldProductB = [120, 140, 160, 190]

#calculate standard deviation for product B units sold
standardDevProductB = np.std(unitsSoldProductB)

print(f'Standard Deviation for Product B units sold is : {standardDevProductB}')

Standard Deviation for Product B units sold is : 25.8602010819715


In [ ]:
#14. Determine the standard deviation of revenue for Product D.
import numpy as np
revenueProductD = [2200, 2700, 3100, 3500]

#calculate standard deveiation for product B revenue
revenueStandardDevProductD = np.std(revenueProductD)

print(f'Standard Deviation for Product D is: {revenueStandardDevProductD}')


Standard Deviation for Product D is: 481.5340071064556


In [10]:
#15. What is the standard deviation of customer satisfaction ratings across all products?
import numpy as np
customerSat = [8, 7, 9, 6, 10]

#calculate customer satisfaction variance
customerSatStandardDev = np.var(customerSat)

print(f'Customer satisfaction rating variance is : {customerSatStandardDev}')

Customer satisfaction rating variance is : 2.0


#### Sampling Questions

In [11]:
#Random Sampling: 
#16. If you randomly select 2 products from the dataset
# what is the probability of selecting a product with a customer satisfaction rating of 9 or above? 

from itertools import combinations

# Customer satisfaction ratings
customerSatisfaction = [8, 7, 9, 6, 10]

# Count products with ratings of 9 or above
products_above_9 = sum(1 for rating in customerSatisfaction if rating >= 9)
total_products = len(customerSatisfaction)

# Total ways to choose 2 products from the total
total_combinations = list(combinations(customerSatisfaction, 2))

# Count combinations with at least one product with a rating of 9 or above
valid_combinations = sum(1 for combo in total_combinations if any(rating >= 9 for rating in combo))

# Calculate probability
probability_at_least_one = valid_combinations / len(total_combinations)

print("Probability of selecting at least one product with a rating of 9 or above:", probability_at_least_one)


Probability of selecting at least one product with a rating of 9 or above: 0.7
